src/class-runner.php

```diff
@@ -0,0 +1,288 @@
+
+		return $processor;
+	}
+
+	/**
+	 * Run a scheduled feed by ID.
+	 *
+	 * @param int $feed_id Feed ID.
+	 * @return void
+	 */
+	public static function run_scheduled( int $feed_id ): void {
+		( new static(
+			$feed_id,
+			function_exists( 'ai_logger_to_post' ) ? ai_logger_to_post( $feed_id, static::LOG_META_KEY, Logger::INFO ) : null,
+		) )->run();
+	}
+
+	/**
+	 * Schedule the next run of a feed.
+	 *
+	 * @param int $feed_id Feed ID.
+	 * @return int|null Timestamp of the next run.
+	 */
+	public static function schedule_next_run( int $feed_id ): ?int {
+		$next_run = wp_next_scheduled( static::CRON_HOOK, [ $feed_id ] );
+
+		if ( $next_run ) {
+			return $next_run;
+		}
+
+		// Determine if the next run should be scheduled.
+		if ( Settings::POST_TYPE !== get_post_type( $feed_id ) || 'publish' !== get_post_status( $feed_id ) ) {
+			return null;
+		}
+
+		// Fetch the frequency of the processor to calculate the next timestamp.
+		$timestamp = time() + static::processor( $feed_id )->frequency();
+
+		if ( ! wp_schedule_single_event( $timestamp, static::CRON_HOOK, [ $feed_id ] ) ) {
+			return null;
+		}
+
+		return $timestamp;
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @param integer              $feed_id Feed post ID.
+	 * @param LoggerInterface|null $logger Logger instance.
+	 */
+	public function __construct( protected int $feed_id, protected ?LoggerInterface $logger = null ) {
+	}
+
+	/**
+	 * Run a feed with the configured settings.
+	 */
+	public function run() {
+		$feed = get_post( $this->feed_id );
+
+		if ( empty( $feed ) ) {
+			$this->logger?->error( 'Feed not found' );
+			return;
+		}
+
+		// Ensure the feed is published.
+		if ( 'publish' !== $feed->post_status ) {
+			$this->logger?->error( 'Feed not published' );
+			return;
+		}
+
+		// Instantiate the processor instance.
+		try {
+			$processor = static::processor( $this->feed_id );
+
+			static::$current_feed_id = $this->feed_id;
+		} catch ( Throwable $e ) {
+			$this->logger?->error( 'Invalid processor', [ 'exception' => $e ] );
+
+			static::$current_feed_id = null;
+
+			return;
+		}
+
+		// Track to New Relic if configured.
+		if ( extension_loaded( 'newrelic' ) ) {
+			if ( function_exists( 'newrelic_name_transaction' ) ) {
+				newrelic_name_transaction( 'feed-consumer' );
+			}
+
+			if ( function_exists( 'newrelic_add_custom_parameter' ) ) {
+				newrelic_add_custom_parameter( 'feed_id', $this->feed_id );
+			}
+		}
+
+		// Clear the previous meta log.
+		delete_post_meta( $this->feed_id, static::LOG_META_KEY );
+
+		$this->logger?->info( 'Run started.' );
+
+		try {
+			$extractor = $processor->extractor();
+
+			$extractor->processor( $processor );
+
+			$extractor->run();
+		} catch ( Throwable $e ) {
+			$this->logger?->error( 'Error running feed extractor', [ 'exception' => $e ] );
+
+			// Schedule the next run of the feed to try again.
+			static::schedule_next_run( $this->feed_id );
```

Comment:
Are there any error states where we wouldn't want to schedule the next run?

Reply:
Good point, it might be nice to back-off and delay the next run but with configuration. Going to ticket.